diff -Naur a/libiberty/cp-demangle.c b/libiberty/cp-demangle.c
--- a/libiberty/cp-demangle.c	2018-10-16 09:01:07.899390967 -0700
+++ b/libiberty/cp-demangle.c	2018-10-16 10:48:55.203283859 -0700
@@ -1758,8 +1758,8 @@
 
 #define NL(s) s, (sizeof s) - 1
 
-CP_STATIC_IF_GLIBCPP_V3
-const struct demangle_operator_info cplus_demangle_operators[] =
+static
+const struct demangle_operator_info cplus_demangle_operators[]  =
 {
   { "aN", NL ("&="),        2 },
   { "aS", NL ("="),         2 },
@@ -1828,9 +1828,14 @@
   { "sz", NL ("sizeof "),   1 },
   { "tr", NL ("throw"),     0 },
   { "tw", NL ("throw "),    1 },
-  { NULL, NULL, 0,          0 }
+  { "",   "",   0,          0 }
 };
 
+__attribute__((visibility("hidden"))) CP_STATIC_IF_GLIBCPP_V3 const struct demangle_operator_info *cplus_get_demangle_operators(void)
+{
+  return cplus_demangle_operators;
+}
+
 static struct demangle_component *
 d_operator_name (struct d_info *di)
 {
@@ -2298,9 +2303,9 @@
                   ::= u <source-name>
 */
 
-CP_STATIC_IF_GLIBCPP_V3
+static
 const struct demangle_builtin_type_info
-cplus_demangle_builtin_types[D_BUILTIN_TYPE_COUNT] =
+cplus_demangle_builtin_types[D_BUILTIN_TYPE_COUNT]  =
 {
   /* a */ { NL ("signed char"),	NL ("signed char"),	D_PRINT_DEFAULT },
   /* b */ { NL ("bool"),	NL ("boolean"),		D_PRINT_BOOL },
@@ -2317,18 +2317,18 @@
   /* h */ { NL ("unsigned char"), NL ("unsigned char"),	D_PRINT_DEFAULT },
   /* i */ { NL ("int"),		NL ("int"),		D_PRINT_INT },
   /* j */ { NL ("unsigned int"), NL ("unsigned"),	D_PRINT_UNSIGNED },
-  /* k */ { NULL, 0,		NULL, 0,		D_PRINT_DEFAULT },
+  /* k */ { ""  , 0,		""  , 0,		D_PRINT_DEFAULT },
   /* l */ { NL ("long"),	NL ("long"),		D_PRINT_LONG },
   /* m */ { NL ("unsigned long"), NL ("unsigned long"),	D_PRINT_UNSIGNED_LONG },
   /* n */ { NL ("__int128"),	NL ("__int128"),	D_PRINT_DEFAULT },
   /* o */ { NL ("unsigned __int128"), NL ("unsigned __int128"),
 	    D_PRINT_DEFAULT },
-  /* p */ { NULL, 0,		NULL, 0,		D_PRINT_DEFAULT },
-  /* q */ { NULL, 0,		NULL, 0,		D_PRINT_DEFAULT },
-  /* r */ { NULL, 0,		NULL, 0,		D_PRINT_DEFAULT },
+  /* p */ { ""  , 0,		""  , 0,		D_PRINT_DEFAULT },
+  /* q */ { ""  , 0,		""  , 0,		D_PRINT_DEFAULT },
+  /* r */ { ""  , 0,		""  , 0,		D_PRINT_DEFAULT },
   /* s */ { NL ("short"),	NL ("short"),		D_PRINT_DEFAULT },
   /* t */ { NL ("unsigned short"), NL ("unsigned short"), D_PRINT_DEFAULT },
-  /* u */ { NULL, 0,		NULL, 0,		D_PRINT_DEFAULT },
+  /* u */ { ""  , 0,		""  , 0,		D_PRINT_DEFAULT },
   /* v */ { NL ("void"),	NL ("void"),		D_PRINT_VOID },
   /* w */ { NL ("wchar_t"),	NL ("char"),		D_PRINT_DEFAULT },
   /* x */ { NL ("long long"),	NL ("long"),		D_PRINT_LONG_LONG },
@@ -2340,6 +2345,11 @@
 	     D_PRINT_DEFAULT },
 };
 
+ __attribute__((visibility("hidden"))) CP_STATIC_IF_GLIBCPP_V3 const struct demangle_builtin_type_info *cplus_get_demangle_builtin_types(void)
+{
+  return cplus_demangle_builtin_types;
+}
+
 CP_STATIC_IF_GLIBCPP_V3
 struct demangle_component *
 cplus_demangle_type (struct d_info *di)
diff -Naur a/libiberty/cp-demangle.h b/libiberty/cp-demangle.h
--- a/libiberty/cp-demangle.h	2018-10-16 09:01:07.903398733 -0700
+++ b/libiberty/cp-demangle.h	2018-10-16 10:49:49.767524761 -0700
@@ -36,9 +36,9 @@
 struct demangle_operator_info
 {
   /* Mangled name.  */
-  const char *code;
+  const char code[3];
   /* Real name.  */
-  const char *name;
+  const char name[17];
   /* Length of real name.  */
   int len;
   /* Number of arguments.  */
@@ -76,11 +76,11 @@
 struct demangle_builtin_type_info
 {
   /* Type name.  */
-  const char *name;
+  const char name[20];
   /* Length of type name.  */
   int len;
   /* Type name when using Java.  */
-  const char *java_name;
+  const char java_name[20];
   /* Length of java name.  */
   int java_len;
   /* How to print a value of this type.  */
@@ -173,15 +173,11 @@
 #define CP_STATIC_IF_GLIBCPP_V3 extern
 #endif
 
-#ifndef IN_GLIBCPP_V3
-extern const struct demangle_operator_info cplus_demangle_operators[];
-#endif
-
 #define D_BUILTIN_TYPE_COUNT (33)
 
-CP_STATIC_IF_GLIBCPP_V3
-const struct demangle_builtin_type_info
-cplus_demangle_builtin_types[D_BUILTIN_TYPE_COUNT];
+CP_STATIC_IF_GLIBCPP_V3 const struct demangle_operator_info *cplus_get_demangle_operators(void);
+
+CP_STATIC_IF_GLIBCPP_V3 const struct demangle_builtin_type_info *cplus_get_demangle_builtin_types(void);
 
 CP_STATIC_IF_GLIBCPP_V3
 struct demangle_component *
diff -Naur a/libiberty/cp-demint.c b/libiberty/cp-demint.c
--- a/libiberty/cp-demint.c	2018-10-16 09:01:07.823243410 -0700
+++ b/libiberty/cp-demint.c	2018-10-16 10:48:12.295455382 -0700
@@ -137,6 +137,9 @@
 {
   int len;
   unsigned int i;
+  const struct demangle_builtin_type_info *cplus_demangle_builtin_types;
+  
+  cplus_demangle_builtin_types = cplus_get_demangle_builtin_types();
 
   if (p == NULL || type_name == NULL)
     return 0;
@@ -163,6 +166,9 @@
 {
   int len;
   unsigned int i;
+  const struct demangle_operator_info *cplus_demangle_operators;
+  
+  cplus_demangle_operators = cplus_get_demangle_operators();
 
   if (p == NULL || opname == NULL)
     return 0;
