From 0043b8832765a146528b4b5660f6082754dba25c Mon Sep 17 00:00:00 2001
From: Azure Sphere Team <azuresphereoss@microsoft.com>
Date: Mon, 11 Dec 2017 13:34:59 -0800
Subject: [PATCH 3/3] Only fetch values for children actually requested

---
 gdb/varobj.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/gdb/varobj.c b/gdb/varobj.c
index 6e8f189..73a3749 100644
--- a/gdb/varobj.c
+++ b/gdb/varobj.c
@@ -867,7 +869,9 @@ varobj_list_children (struct varobj *var, int *from, int *to)
   while (var->children.size () < var->num_children)
     var->children.push_back (NULL);
 
-  for (int i = 0; i < var->num_children; i++)
+  varobj_restrict_range(var->children, from, to);
+
+  for (int i = *from; i < *to; i++)
     {
       if (var->children[i] == NULL)
 	{
@@ -880,7 +880,6 @@ varobj_list_children (struct varobj *var, int *from, int *to)
 	}
     }
 
-  varobj_restrict_range (var->children, from, to);
   return var->children;
 }
 
-- 
2.7.4

